src/video_canister/src/lib.rs

```diff
@@ -0,0 +1,148 @@
+    pub chunk_num: usize,
+    pub version: usize,
+    pub owner: Principal,
+}
+
+impl Default for MetaInfo{
+    fn default() -> Self { 
+        MetaInfo{
+            name: String::new(),
+            description: String::new(),
+            chunk_num: 0,
+            version: CANISTER_VERSION,
+            owner: Principal::anonymous(),
+        }
+     }
+}
+
+#[derive(CandidType, Deserialize)]
+pub enum PutMetaInfoResponse{
+    #[serde(rename = "success")]
+    Success,
+
+    #[serde(rename = "missing_rights")]
+    MissingRights,
+}
+
+#[derive(CandidType, Deserialize)]
+pub enum PutChunkResponse{
+    #[serde(rename = "success")]
+    Success,
+
+    #[serde(rename = "missing_rights")]
+    MissingRights,
+
+    #[serde(rename = "out_of_bounds")]
+    OutOfBounds
+}
+
+#[derive(CandidType, Deserialize)]
+pub enum ChangeOwnerResponse{
+    #[serde(rename = "success")]
+    Success,
+
+    #[serde(rename = "missing_rights")]
+    MissingRights,
+}
+
+type Chunk = Vec<u8>;
+type Chunks = Vec<Chunk>;
+
+const CANISTER_VERSION: usize = 0usize;
+
+#[init]
+pub async fn init(owner: Principal){
+    storage::get_mut::<MetaInfo>().owner = owner;
+}
+
+#[update]
+pub async fn put_meta_info(new_meta_info: PutMetaInfo) -> PutMetaInfoResponse{
+    let mut old_meta_info = storage::get_mut::<MetaInfo>();
+    if old_meta_info.owner != ic_cdk::caller() {
+        return PutMetaInfoResponse::MissingRights;
+    } else {
+        old_meta_info.name = new_meta_info.name;
+        old_meta_info.description = new_meta_info.description;
+
+        //Resize array
+        if old_meta_info.chunk_num != new_meta_info.chunk_num{
+            let _old_chunks = std::mem::replace(storage::get_mut::<Chunks>(), vec![Vec::new(); new_meta_info.chunk_num]);
+            old_meta_info.chunk_num = new_meta_info.chunk_num;
+        }
+
+        return PutMetaInfoResponse::Success;
+    }
+}
+
+#[update]
+pub async fn put_chunk(chunk_num: usize, chunk: Chunk) -> PutChunkResponse{
+    let owner = (*storage::get::<MetaInfo>()).owner;
+    if owner != ic_cdk::caller() {
+        return PutChunkResponse::MissingRights;
+    } else {
+        let chunks = storage::get_mut::<Chunks>();
+
+        if chunk_num >= chunks.len(){
+            return PutChunkResponse::OutOfBounds;
+        }
+        
+        chunks[chunk_num] = chunk;
+        return PutChunkResponse::Success;
+    }
+}
+
+#[update]
+pub async fn change_owner(new_owner: Principal) -> ChangeOwnerResponse{
+    let old_owner = &mut storage::get_mut::<MetaInfo>().owner;
+
+    if *old_owner != ic_cdk::caller() {
+        return ChangeOwnerResponse::MissingRights;
+    } else { 
+        *old_owner = new_owner;
+        return ChangeOwnerResponse::Success;
+    }
+}
+
+#[query]
+pub async fn get_meta_info() -> &'static MetaInfo{
+    return storage::get::<MetaInfo>();
+}
+
+#[query]
+pub async fn get_chunk(chunk_num: usize) -> Option<&'static Chunk>{
+    return storage::get::<Chunks>().get(chunk_num);
```

Comment:
"Returns a share reference of the storage. This will create the storage if it doesn’t exist."
Source: https://docs.rs/ic-cdk/latest/ic_cdk/storage/fn.get.html

Dementsprechend sollte davor ein check kommen ob chunk_num < Chunks.len, no?

Reply:
Nein, da wird zuerst der Chunk vector geholt und dann der dementsprechende chunk aus dem **vector** gezogen, exisitiert der nicht dann wird None returned siehe https://doc.rust-lang.org/std/vec/struct.Vec.html#method.get

Klarer implementiert in b1f23219ed87869dc1f408661ff5defda87ece02